File: lab2-functionalprimer/python3/func_as_arg.py
def listFunc(a, b):
    return [i for i in range(a, b + 1)]  


def applicatorFunc(inpFunc, a, b, s):
    numbers = inpFunc(a, b)  
    if s == 's':
        return sum(numbers)  
    elif s == 'a':
        return sum(numbers) / len(numbers)  
    else:
        return "Invalid option! Use 's' for sum or 'a' for average."


a = int(input("Enter the starting number (a): "))
b = int(input("Enter the ending number (b): "))
s = input("Enter 's' for sum or 'a' for average: ")


print("Result =", applicatorFunc(listFunc, a, b, s))

File: lab2-functionalprimer/haskell/func_as_arg.hs
inpFunc :: Int -> Int -> [Int]
inpFunc a b = [a..b]


applicatorFunc :: [Int] -> Char -> Double
applicatorFunc inp s
    | s == 's'  = fromIntegral (sum inp)
    | s == 'a'  = fromIntegral (sum inp) / fromIntegral (length inp)  
    | otherwise = error "Invalid option! Please enter 's' for sum or 'a' for average."


main = do
    putStrLn "Enter the starting number (a):"
    a <- readLn
    putStrLn "Enter the ending number (b):"
    b <- readLn
    putStrLn "Enter 's' for sum or 'a' for average:"
    op <- getLine
    let result = applicatorFunc (inpFunc a b) (head op)  
    putStrLn ("Result = " ++ show result)
